Improve docstrings for `Int`, `Float64`, `^`, and friends

mcabbott opened the pull request:
This extends the docs for `Int`, `Float64` etc, to
* Note that Int64 is the default, mention that it can overflow
* Explain what `1f0` means, and that Float64 is the default

Similarly extends docs for `^`, `+`, `*` aiming to
* Point out that `1.2 * 10^3` is a bad habit, warn about overflow
* Give an example for what `literal_pow` is doing since the explanation is quite technical
* Also point out overflow in `+`, as [suggested here](https://github.com/JuliaLang/julia/pull/45141#discussion_r862808888)
* While there, mention that you can add vectors, and that vararg `+(1,2,3,4)` has a default binary behaviour
* Similarly mention that vararg `*(1,2,3,4)` has a default order, from the left
* While there, mention `1/2pi` and `v'v` as examples of `*`.

mcabbott committed:
FloatN docs

mcabbott committed:
FloatN docs

mcabbott committed:
integer docs

mcabbott committed:
integer docs

mcabbott committed:
mention overflow in docs for +, ^.

mcabbott committed:
mention overflow in docs for +, ^.

mcabbott committed:
update docs for * to match

mcabbott committed:
update docs for * to match

mcabbott commented on base/promotion.jl:
```diff
@@ -403,20 +407,34 @@ where usually `^ == Base.^` unless `^` has been defined in the calling
+julia> big(10)^19 == 1e19
+true
 
-julia> A^3
-2×2 Matrix{Int64}:
- 37   54
- 81  118
```
The method for matrix powers has its own docstring, with examples.

LilithHafner replied:
Good catch! I dislike it when docstrings get too long with redundant examples across many methods. I often just want the first sentence or paragraph and the first thing I see at the REPL is a long list of examples for some esoteric method defined in a package.

mcabbott replied:
Am certainly guilty of extending quite a few... It's awkward because well-chosen examples, graded from simple to complicated, are actually very helpful. But scrolling past ones you don't care about is a pain. We have `??` for "extended help" but this doesn't seem like a great solution.

mcabbott commented on base/docs/basedocs.jl:
````diff
@@ -1783,6 +1783,18 @@ Number
     Real <: Number
 
 Abstract supertype for all real numbers.
+
+```jldoctest
+julia> subtypes(Real)
+4-element Vector{Any}:
+ AbstractFloat
+ AbstractIrrational
+ Integer
+ Rational
````
This fails on CI, not sure why.

Help for `subtypes` has a similar example, is this one being processed earlier in loading or something?

mcabbott commented on base/docs/basedocs.jl:
```diff
@@ -1841,32 +1888,67 @@ See also: [`digits`](@ref), [`iszero`](@ref), [`NaN`](@ref).
-        """
-            Float$($bit) <: AbstractFloat
+"""
+    Float64 <: AbstractFloat <: Real
```
The reason to add `<: Real` is seeing one too many functions typed `f(x::AbstractFloat)`.

mcabbott committed:
complaints from CI

mcabbott committed:
complaints from CI

mcabbott committed:
significand, exponent and friends

mcabbott committed:
significand, exponent and friends

KristofferC commented on base/docs/basedocs.jl:
```diff
@@ -1811,6 +1825,21 @@ Signed
 
 Abstract supertype for all unsigned integers.
+
+All are printed in hexadecimal, with prefix `0x`, 
```
```suggestion
All unsigned integers are printed in hexadecimal, with prefix `0x`, 
```

I guess?

KristofferC commented on base/docs/basedocs.jl:
````diff
@@ -1811,6 +1825,21 @@ Signed
+
+# Examples
+```
+julia> unsigned(true)
````
Feels like an odd example to have as the first one. Imo the examples should go from the most obvious to the more "esoteric"

mcabbott replied:
Yes, good point.

mcabbott commented on base/docs/basedocs.jl:
````diff
@@ -1797,6 +1797,20 @@ AbstractFloat
+
+# Examples
+```
+julia> 42 isa Integer
+true
+
+julia> 1.0 isa Integer
````
Maybe this should mention Int, somehow:
```suggestion
julia> 42 isa Integer  # concrete type Int
true

julia> 1.0 isa Integer  # concrete type Float64
```

halleysfifthinc commented on base/docs/basedocs.jl:
```diff
@@ -1811,6 +1825,21 @@ Signed
 
 Abstract supertype for all unsigned integers.
+
+All unsigned integers are printed in hexadecimal, with prefix `0x`, 
+and can be entered in the same way.
```
```suggestion
All unsigned integers are printed and can be parsed in hexadecimal, with prefix `0x`.
```

mcabbott replied:
Can you say why?

I guess I basically want to say you can type that in, but type with a keyboard is unclear when there are data types around. But it's a minor point since by default most printing can be pasted back in. That anything you type in is parsed seems like something people who know know, and who don't, don't need to.

halleysfifthinc replied:
Ah! I missed the potential ambiguity of `type`/typing. "entered" reads awkwardly to me. Perhaps "written" instead?

Either 
```suggestion
All unsigned integers are printed in hexadecimal, with prefix `0x`, 
and can be written in the same way.
```
or
```suggestion
All unsigned integers are printed and can be written in hexadecimal, with prefix `0x`.
```

mcabbott commented on base/docs/basedocs.jl:
```diff
@@ -1841,32 +1870,69 @@ See also: [`digits`](@ref), [`iszero`](@ref), [`NaN`](@ref).
+"""
+Float16
 
 for bit in (8, 16, 32, 64, 128)
+    type = Symbol(:Int, bit)
+    default = eval(type) == Int ? "\n \n This is the default used for most integer literals (on $bit systems), \
+    and has an alias `Int === $type`" : ""
```
Instead of `eval` this can use `Sys.WORD_SIZE`. And should perhaps mention that? Explain why there's no corresponding `Float`?

mcabbott replied:
Instead of `eval` this can use `Sys.WORD_SIZE`. And should perhaps mention that? Explain why there's no corresponding `Float`?

mcabbott committed:
tweak

mcabbott committed:
tweak

mcabbott commented:
Bump?

Related PRs (after this one) include #45975 (type trees for abstract types) #45415 (Int and UInt, in addition to Int64 etc.)

mcabbott committed:
move Int and UInt out of eval loop

mcabbott committed:
move Int and UInt out of eval loop

mcabbott committed:
attach to Int64 etc, to please CI?

mcabbott committed:
attach to Int64 etc, to please CI?

mcabbott committed:
upgrades to Bool, why not

mcabbott committed:
upgrades to Bool, why not

mcabbott committed:
whitespace, etc

mcabbott committed:
whitespace, etc

mcabbott commented on base/docs/basedocs.jl:
```diff
@@ -1977,57 +2006,146 @@ Boolean type, containing the values `true` and `false`.
+    """
+    $(Symbol(Int))
+
+    """
+        $UInt === UInt
+
+    $(Sys.WORD_SIZE)-bit unsigned integer type, `$UInt <: Unsigned <: Integer`.
+
+    Like [`Int`](@ref $Int), the alias `UInt` may point to either `UInt32` or `UInt64`,
+    according to the value of `Sys.WORD_SIZE` on a given computer.
```
As suggested by [jakobnissen](https://github.com/jakobnissen), this adds a note that `UInt64 == UInt` etc. 

The docstring for `Int64` has also had more detail added, to indicate how widely the default is used.

mcabbott commented on base/docs/basedocs.jl:
````diff
@@ -1977,57 +2006,146 @@ Boolean type, containing the values `true` and `false`.
+```jldoctest
+julia> [1 2 3 4 5] .< pi
+1×5 BitMatrix:
+ 1  1  1  0  0
````
The other new change is to expand on `Bool` a bit, also an integer. 

One goal of changes here is to add a reminder that the printing of arrays of Bool may use `0, 1`.

Another is to note that "strong zero" covers Inf as well as NaN.

mcabbott committed:
fix a ref to Int

mcabbott committed:
fix a ref to Int

mcabbott commented:
Bump?

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -1977,57 +2006,146 @@ Boolean type, containing the values `true` and `false`.
 
         $($bit)-bit signed integer type.
+
+        Note that integers overflow without warning,
+        thus `typemax($($type)) + $($type)(1) < 0`.
```
So do unsigned integers. I think it could be misleading to add this note only to signed BitIntegers.

mcabbott replied:
Is "bitinteger" a word? The comment is specifically about `Int16` or whatever, perhaps shouldn't read as applying to all integers:
```suggestion
        Note that such integers overflow without warning,
        thus `typemax($($type)) + $($type)(1) < 0`.
```

mcabbott replied:
And, a similar comment could apply to unsigned. Somehow overflow seems much less surprising there, as they are printed in hex and obviously a representation of some bits, not an integer like when you learned to count.

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -1967,6 +1981,21 @@ Signed
 
 Abstract supertype for all unsigned integers.
+
+All are printed in hexadecimal, with prefix `0x`,
+and can be entered in the same way.
```
This is true for all subtypes of `Unsigned` defined in Base but does not necessarily hold for subtypes defined in packages.

mcabbott replied:
It could say something like this:
```suggestion
All built-in unsigned integers printed in hexadecimal, with prefix `0x`,
and can be entered in the same way.
```
Are there any defined elsewhere which don't obey this? Defining a new unsigned integer type sounds like an odd thing to do.

LilithHafner replied:
SaferIntegers.jl defines `SafeUInt8 <: SafeUnsgined <: Unsgined` which errors on overflow. These are printed in hexadecimal, but cannot be entered in the same way
```julia
julia> x = SafeUnsigned(0x17)
0x17

julia> x === 0x17
false
```
Perhaps drop the "all"
```suggestion
Built-in unsigned integers are printed in hexadecimal, with prefix `0x`,
and can be entered in the same way.
```

mcabbott replied:
Done.

LilithHafner commented on base/docs/basedocs.jl:
````diff
@@ -1977,57 +2006,146 @@ Boolean type, containing the values `true` and `false`.
 ```
 
-See also: [`digits`](@ref), [`iszero`](@ref), [`NaN`](@ref).
+Branches via [`if`](@ref) only accept `Bool`, there are no "truthy" values in Julia.
````
```suggestion
Branches via [`if`](@ref) and other conditionals only accept `Bool`.
There are no "truthy" values in Julia.
```
(other examples include `&&`, `||`, `?/:` and `while`)

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -1977,57 +2006,146 @@ Boolean type, containing the values `true` and `false`.
 
+@eval begin
+    """
+        $Int === Int
```
This is works locally, but IIUC will result in the web-hosted documentation listing `Int64 === Int` which is not good.

mcabbott replied:
OK, that's awkward, but what should it say?

mcabbott replied:
Maybe for now we remove it. 

Can we just re-word below "This is the default type of most integer literals" to indicate that it means "on 64-bit systems"?
```suggestion
        $Int
```

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -1977,57 +2006,146 @@ Boolean type, containing the values `true` and `false`.
+    The default (and the alias) may be 32 or 64 bits on different computers, indicated by `Sys.WORD_SIZE`.
+
+    Note that integers overflow without warning, thus `typemax(Int) + 1 < 0` and `10^19 < 0`.
+    Overflow can be avoided by using [`BigInt`](@ref), and is checked by most operations on [`Rational`](@ref).
```
Most? I thought all rational operations were checked, but I'm not particularly confident.

mcabbott replied:
Likewise! Even if all in Base are, some package may define operations which aren't.

LilithHafner replied:
On second thought, I don't think this statement about rationals belongs in the Int docstring. At most a "see also Rational" but even that seems like a bit of a stretch.

mcabbott replied:
Could be removed:
```suggestion
    Overflow can be avoided by using [`BigInt`](@ref).
```

mcabbott replied:
Done.

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -2568,15 +2709,28 @@ julia> 2 * 7 * 8
+  2
+ 30
+
+julia> 1/2pi, 1/2*pi  # juxtaposition has higher precedence
+(0.15915494309189535, 1.5707963267948966)
+
+julia> x = [1, 2]; x'x  # adjoint vector * vector
+5
```
Probably a good idea to mention juxtaposition in the second paragraph too.

mcabbott replied:
Now done.

LilithHafner commented on base/promotion.jl:
```diff
@@ -415,7 +415,11 @@ end
+
+If `x` and `y` are integers, the result may overflow.
+To enter numbers in scientific notation, use [`Float64`](@ref) literals
+such as `1.2e3` rather than `1.2 * 10^3`.
```
For literals, if we don't change the current behavior, this seems like something a linter should catch, but that's not within the scope of the JuliaLang/julia repo.

LilithHafner reviewed (changes requested):
Thank you for the great work @mcabbott! These are important documentation improvements, many of which look good to me. Nevertheless, there are still a few changes I think need to be made before merging.

I think we need a cohesive story for how we warn about overflow. Currently, this PR adds a warning for IntXX but not UIntXX, and for + and ^ but not *.

Generating different docstrings for different Sys.WORD_SIZE might be acceptable, but we would need a third branch for when we are building the web-hosted documentation. Even then, it could get dicey as someone developing on a 64 bit version of Julia might glance at the top of the docstring only and falsely find that `Int64 === Int` holds unconditionally.

mcabbott commented on base/docs/basedocs.jl:
```diff
@@ -1977,57 +2006,146 @@ Boolean type, containing the values `true` and `false`.
+Unlike integers, this default does not change with `Sys.WORD_SIZE`.
+
+The exponent for scientific notation can be entered as `e` or `E`,
+thus `2e3 === 2.0E3 === 2.0 * 10^3`. This is strongly preferred over
+`10^n` because integers overflow, thus `10^19 < 0`.
```
This paragraph is about entering floats. There, the preference is very strong, no? Writing `1.5 * 10^12` to enter a number is simply a mistake.

mcabbott replied:
```suggestion
thus `2e3 === 2.0E3 === 2.0 * 10^3`. Doing so is strongly preferred over
`10^n` because integers overflow, thus `2.0 * 10^19 < 0` but `2e19 > 0`.
```

Albert55554 commented:
--Sent from my Android phone with mail.com Mail. Please excuse my brevity.

mcabbott commented on base/docs/basedocs.jl:
```diff
@@ -2559,7 +2697,10 @@ julia> -(2, 4.5)
+Multiplication operator.
+
+Infix `x*y*z*...` calls this function with all arguments, i.e. `*(x, y, z, ...)`,
+which by default then calls `(x*y) * z * ...` starting from the left.
```
As suggested below, not sure about wording:
```suggestion
which by default then calls `(x*y) * z * ...` starting from the left.

Juxtaposition such as `2pi` also calls `*(2, pi)`. Note that this is parsed with higher precedence.
```

mcabbott commented on base/docs/basedocs.jl:
````diff
@@ -1977,57 +2006,146 @@ Boolean type, containing the values `true` and `false`.
+ 0  0  0  1  1
+```
+
+See also [`trues`](@ref), [`falses`](@ref), [`digits`](@ref), [`ifelse`](@ref).
````
It's not the focus of `digits`, but more something you might be looking for? Not at all set on it.

mcabbott committed:
Apply suggestions from code review

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

LilithHafner commented:
@Albert55554, your comment appears to be entirely empty.

mcabbott committed:
Merge branch 'doc6' of github.com:mcabbott/julia into doc6

mcabbott committed:
apply suggestions, change wording for Int64

mcabbott commented:
> Currently, this PR adds a warning for IntXX but not UIntXX, 

There's some logic to this. You may expect `Int` to just work like integers in mathematics, they are printed as just numbers, this is an easy mistake to make. But if you are using `UInt8`, then you are choosing to work with bit patterns, and obviously expected to know more.

> and for + and ^ but not *.

I think this was because that's where it came up. But I've added a similar sentence to `*`. 

> for when we are building the web-hosted documentation

I've removed the claim `Int64 === Int`, and mention the alias only in words, after mentioning `Sys.WORD_SIZE`. Not sure the wording is ideal but see what you think. 

Am I correct in thinking that working on a 32-bit system is now something very unusual? The web documentation must not lie, but it does not have to be completely even-handed. Trying to generate a 3rd version for this sounds like a lot of work and a lot of opportunities for bugs, for little gain.

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -1977,57 +2006,151 @@ Boolean type, containing the values `true` and `false`.
+    Like [`Int`](@ref $Int), the alias `UInt` may point to either `UInt32` or `UInt64`,
+    according to the value of `Sys.WORD_SIZE` on a given computer.
+
+    Printed in hexadecimal, thus `UInt(15) === $(repr(UInt(15)))`.
```
This holds because they are parsed in hexadecimal, not because they are printed in hexadecimal.

mcabbott replied:
```suggestion
    Printed and parsed in hexadecimal: `UInt(15) === $(repr(UInt(15)))`.
```

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -2568,15 +2721,28 @@ julia> 2 * 7 * 8
+  2
+ 30
+
+julia> 1/2pi, 1/2*pi  # juxtaposition has higher precedence
+(0.15915494309189535, 1.5707963267948966)
+
```
This example is already listed in the text above.

mcabbott replied:
Well, the other way around. Maybe the text is too long?

mcabbott commented on base/docs/basedocs.jl:
```diff
@@ -2559,7 +2702,17 @@ julia> -(2, 4.5)
+Infix `x*y*z*...` calls this function with all arguments, i.e. `*(x, y, z, ...)`,
+which by default then calls `(x*y) * z * ...` starting from the left.
+
+Juxtaposition such as `2pi` also calls `*(2, pi)`. Note that this operation
+has higher precedence than a literal `*`, so that for instance `1/2pi == 1/(2 * pi)`,
+but lower precedence than `^`, so that `5x^2 == 5 * (x^2)`.
```
Paragraph added in response to a suggestion above. But perhaps too much overlap with the examples. And perhaps better to say other things?
```suggestion
Juxtaposition such as `2pi` also calls `*(2, pi)`. Note that this operation
has higher precedence than a literal `*`. Note also that juxtaposition "0x..."
(integer zero times a variable whose name starts with `x`) is forbidden as
it clashes with unsigned integers literals: `0x01 isa UInt8`.
```

mcabbott committed:
Update base/docs/basedocs.jl

mcabbott committed:
Update base/docs/basedocs.jl

mcabbott committed:
Merge branch 'master' into doc6

mcabbott commented:
Bump, maybe it's worth trying to finish this?

Today's question about `10^23` is: https://discourse.julialang.org/t/scientific-notation-vs-10/98849

fingolfin commented on base/docs/basedocs.jl:
```diff
@@ -2028,57 +2057,151 @@ Boolean type, containing the values `true` and `false`.
+thus `2e3 === 2.0E3 === 2.0 * 10^3`. Doing so is strongly preferred over
+`10^n` because integers overflow, thus `2.0 * 10^19 < 0` but `2e19 > 0`.
+
+See also [`Inf`](@ref), [`NaN`](@ref), [`floatmax`](@ref), [`Float32`](@ref), [`Complex`](@ref).
```
These "See also" lists seem somewhat arbitrary?
- `Float16` has no "See also" at all (it could be point to e.g. `Inf16`, `NaN16`, ...)
- here we point to `floatmax` and `Complex` but Float32 does not.
- on the other hand, Float32 points to `exponent` and `frexp`?

What is the logic behind this?

mcabbott replied:
They are all a bit random, not just here. Maybe the pattern is:
* link to obvious variants, like `sort` vs `sort!`, but not too many, no complete table UInt8 to Int256.
* introduce a few neat things, which you guess might be of interest. The more obscure the function, the more you should assume the reader knows the basics. Sometimes there are many candidates, better to pick different subsets each time than consistently the same.

mcabbott replied:
> Float32 points to exponent and frexp

These are discussed earlier in the Float64 text.

> Float16 ... could be point to e.g. Inf16, NaN16, ...

Sure. It's also much more niche than the others. Maybe readers have a firm grasp of Float32 already.

vtjnash commented:
bump. Can we finish or close whatever is wanted here?

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -2003,7 +2003,21 @@ AbstractFloat
     Integer <: Real
 
-Abstract supertype for all integers.
+Abstract supertype for all integers, [`Signed`](@ref), [`Unsigned`](@ref) or [`Bool`](@ref).
```
```suggestion
Abstract supertype for all integers (e.g. [`Signed`](@ref), [`Unsigned`](@ref), and [`Bool`](@ref)).
```

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -2028,57 +2057,151 @@ Boolean type, containing the values `true` and `false`.
+Branches via [`if`](@ref) and other conditionals only accept `Bool`.
+There are no "truthy" values in Julia.
+
+Comparisons return `Bool`, and broadcasted comparisons may
```
```suggestion
Comparisons typically return `Bool`, and broadcasted comparisons may
```

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -2028,57 +2057,151 @@ Boolean type, containing the values `true` and `false`.
+    """
+        $Int
+
+    $(Sys.WORD_SIZE)-bit signed integer type, `$Int <: Signed <: Integer <: Real`.
```
This interpolation will make the web-hosted documentation falsely state that Int64 "is the type returned by functions such as `length`".

mcabbott replied:
I don't have a great solution here. Would welcome re-wording solutions if you feel it's necessary for the web version never to reflect the machine it was generated on. 32 bit systems are pretty rare, and if you keep reading it may make sense... can we just add a disclaimer?

LilithHafner reviewed (commented):
This is almost ready to merge.

I left a few small comments, but the only merge-blocker is that, as is, this PR will make the web-hosted documentation state that `Int64` is the type returned by `length`

Giving `Int` a docstring separate from `Int64` and `Int32` would be much better.

```julia
"""
   I32

32 bit integer
"""
struct I32 end

"""
   I64

64 bit integer
"""
struct I64 end

"""
    I

an alias for I32 or I64, depending on Sys.WORD_SIZE
"""
const I = Sys.WORD_SIZE == 64 ? I64 : I32
```

```
help?> I32
search: I32 Int32 Inf32 UInt32

  I32

  32 bit integer

help?> I64
search: I64 Int64 Inf64 UInt64 Int16 Inf16

  I64

  64 bit integer

help?> I
search: I IO if Int in im Inf I64 I32 isa Int8 inv Int64 Int32 Int16 imag Inf64 Inf32 Inf16 Int128 isqrt isone isodd isnan isinf isdir IdDict import invoke Integer iszero isreal

  I

  an alias for I32 or I64, depending on Sys.WORD_SIZE
```

mcabbott commented:
> Giving Int a docstring separate from Int64 and Int32 would be much better

It's been a while, but I'm pretty sure I tried this, and was unable to make CI / documenter understand it correctly.

mcabbott committed:
Apply 2 suggestions from code review

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

mcabbott committed:
Merge branch 'master' into doc6

LilithHafner commented:
> It's been a while, but I'm pretty sure I tried this, and after a few days was unable to make CI / documenter understand it correctly. 

If the code in https://github.com/JuliaLang/julia/pull/45221#pullrequestreview-1708676817 doesn't work that feels like a bug/shortcoming of Documenter, but I expect it should work. I'll try to push a commit using that approach.

> Perhaps it can be attempted in a follow-up PR.

I don't want to merge something that knowingly introduces false statements to the documentation, especially this close to the feature freeze.

LilithHafner committed:
Give Int/UInt a separate docstring

LilithHafner commented:
With 7e98348 we have

```julia
help?> Int64
search: Int64 UInt64 Inf64 Int8 Int16 Int32 Int UInt32

  Int64 <: Signed <: Integer

  64-bit signed integer type.

  Note that such integers overflow without warning,
  thus typemax(Int64) + Int64(1) < 0.

  See also Int, widen, BigInt.

help?> Int32
search: Int32 UInt32 Inf32 Int128 Int8 Int16 Int64 Int

  Int32 <: Signed <: Integer

  32-bit signed integer type.

  Note that such integers overflow without warning,
  thus typemax(Int32) + Int32(1) < 0.

  See also Int, widen, BigInt.

help?> Int
search: Int Int8 UInt Inf Int16 Int32 Int64 UInt8

  Int

  Sys.WORD_SIZE-bit signed integer type, Int <:
  Signed <: Integer <: Real.

  This is the default type of most integer literals,
  on computers for which Sys.WORDSIZE == 64, and has
  an alias Int. It is the type returned by functions
  such as length, and the standard type for indexing
  arrays. The default (and the alias) may be 32 or 64
  bits on different computers, indicated by
  Sys.WORD_SIZE.

  Note that integers overflow without warning, thus
  typemax(Int) + 1 < 0 and 10^19 < 0. Overflow can be
  avoided by using BigInt. Very large integer
  literals will use a wider type, for instance
  10_000_000_000_000_000_000 isa Int128.

  Integer division is div alias ÷, whereas / acting
  on integers returns Float64.

  See also Int64, widen, typemax, bitstring.
```

And all three docstrings are listed in the manual under `en/base/numbers.html#Concrete-number-types`. If Documenter.jl and CI are happy with this we can update the Int docstring appropriately.

mcabbott commented:
>  knowingly introduces false statements 

This is a little strong! The suggestion was to water down the text sufficiently (which is easy) then merge. Then fight CI, if so inclined.

If you want to fight Documenter.jl and CI on this branch instead, that's also OK. It's entirely possible I missed some trick when I tried before.

LilithHafner commented:
Ah, I was assuming you were proposing merging as was.

Anyway, Documentation and CI are green after implementing [my previous suggestion](https://github.com/JuliaLang/julia/pull/45221#pullrequestreview-1708676817), no fight required.

mcabbott commented on base/docs/basedocs.jl:
```diff
@@ -2053,57 +2082,148 @@ Boolean type, containing the values `true` and `false`.
+
+The exponent for scientific notation should be entered as lower-case `f`,
+thus `2f3 === 2.0f0 * 10^3 === Float32(2_000)`.
+For arrays literals and comprehensions, the element type can be specified before
```
```suggestion
For array literals and comprehensions, the element type can be specified before
```

LilithHafner committed:
Int docstring wording fixup

mcabbott commented on base/docs/basedocs.jl:
```diff
@@ -2851,7 +2990,18 @@ julia> -(2, 4.5)
+Juxtaposition such as `2pi` also calls `*(2, pi)`. Note that this operation
+has higher precedence than a literal `*`. Note also that juxtaposition "0x..."
+(integer zero times a variable whose name starts with `x`) is forbidden as
+it clashes with unsigned integers literals: `0x01 isa UInt8`.
```
```suggestion
it clashes with unsigned integer literals: `0x01 isa UInt8`.
```

mcabbott committed:
Apply suggestions from code review

LilithHafner committed:
Remove trailing whitespace

LilithHafner commented on base/docs/basedocs.jl:
```diff
@@ -2053,57 +2082,147 @@ Boolean type, containing the values `true` and `false`.
+
+Sys.WORD_SIZE-bit signed integer type, `Int <: Signed <: Integer <: Real`.
+
+This is the default type of most integer literals and is an ailias for either `Int32`
```
```suggestion
This is the default type of most integer literals and is an alias for either `Int32`
```

LilithHafner commented on base/math.jl:
```diff
@@ -1010,6 +1012,8 @@ For any other non-subnormal floating-point number `x`, this corresponds to the e
 
 See also [`signbit`](@ref), [`significand`](@ref), [`frexp`](@ref), [`issubnormal`](@ref), [`log2`](@ref).
 
+See also [`significand`](@ref), [`exponent`](@ref), [`ldexp`](@ref).
+
```
These should be merged

```suggestion
See also [`signbit`](@ref), [`significand`](@ref), [`frexp`](@ref), [`issubnormal`](@ref), [`log2`](@ref), [`ldexp`](@ref).
```

mcabbott replied:
Merge conflict of sorts with #47085

LilithHafner committed:
Apply suggestions from code review

mcabbott commented:
Thanks for pushing this over the finish line!

LilithHafner commented:
Thank you for doing the bulk of the work!
